### PR TITLE
Trip with Fixed Start and End points (TFSE)

### PR DESCRIPTION
**Pull request by ghoshkaj**
# Issue

https://github.com/Project-OSRM/osrm-backend/issues/1623

## Tasklist
 - [x] Work with Chau to figure out algorithm for roundtrip with fixed start and end points 
(We're going to implement @chaupow's algorithm described [here](https://github.com/Project-OSRM/osrm-backend/issues/1623#issuecomment-26510917))
 - [x] Figure out API design 
(@chaupow and @TheMarex and I have settled on: 
`coordinates`: list of coordinates
`source`: the index of the start coordinate
`destination`: the index of the end coordinate )
- [x] Write failing unit tests for new parameters
- [x] Make unit tests pass
- [x] Write failing cuke tests
- [x] Write code
- [ ] Review
- [ ] Adjust for comments

## Code Review Checklist - author check these when done, reviewer verify
 - [x] Code formatted with `scripts/format.sh`
 - [x] Changes have test coverage
 - [ ] New exceptions, logging, errors - are messages distinct enough to track down in the code if they get thrown in production on non-debug builds?
 - [x] The PR is one logically integrated piece of work.  If there are unrelated changes, are they at least separate commits?
 - [x] Commit messages - are they clear enough to understand the intent of the change if we have to look at them later?
 - [x] Code comments - are there comments explaining the intent?
 - [x] Relevant docs updated
 - [x] Changelog entry if required
 - [ ] Impact on the API surface
   - [ ] If HTTP/libosrm.o is backward compatible features, bump the minor version
   - [ ] File format changes require at minor release
   - [ ] If old clients can't use the API after changes, bump the major version

If something doesn't apply, please ~~cross it out~~

## Requirements / Relations
 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by danpat** [comment]
Capturing from chat today:

The complexity seems quite high with the approach of creating the tfse_ table, performing the FA/brute search, then mapping the results back to the original node IDs.

An alternative could be just to munge certain values in the distance table.  If `A` is our start point, and `D` is our designated endpoint, we can do something like set all "X->A" cell values to `INT_MAX`, and all `D->X` values to `INT_MAX`.  This should be basically equivalent to the `tfse_table_`, but without needing to perform that step.

Of course, with any heuristic, it's possible that we'd calculate a result that includes an `INT_MAX` edge, and this could cause all kinds of integer overflow problems.  To make the result correct, you can update the brute force and FA algorithms to fast-fail a candidate route if it touches an `INT_MAX` edge - i.e. `INT_MAX` is treated as a special value that means "no connection".

@chaupow Thoughts on the above?  The creation of the `tfse_table_` makes sense, but it adds some complexity that I think we could avoid.  Can you see anything fundamentally wrong with the approach above?

**Comment by chaupow** [comment]
> @chaupow Thoughts on the above? The creation of the tfse_table_ makes sense, but it adds some complexity that I think we could avoid. Can you see anything fundamentally wrong with the approach above?

If I'm not wrong, this should work as well and result in the same effect of forcing a round trip that visits A after D. Also, Julien from vroom does the same https://github.com/Project-OSRM/osrm-backend/issues/1623#issuecomment-267534420

I'm 👍

**Comment by ghoshkaj** [comment]
This is the table that @danpat and I came up with from our conversation. Putting it here for the record as I work through the cucumber test scenarios that use this and so that everyone is on the same page about new table.

original table:
```  
  a  b  c  d  e
a 0  15 36 34 30 
b 15 0  25 30 34 
c 36 25 0  18 32 
d 34 30 18 0  15 
e 30 34 32 15 0 
```
manipulated table:  
```
  a      b      c      d      e
a 0      15     10000  34     30 
b 10000  0      25     30     34 
c 0      10000  10000  10000  10000 
d 10000  30     18     0      15 
e 10000  34     32     15     0
```

**Comment by ghoshkaj** [comment]
@danpat @chaupow and anyone else on directions, review please?

File: src/engine/plugins/trip.cpp [hunk 1]
```diff
@@ -187,8 +204,90 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+    std::vector<EdgeWeight> tfse_table_ = result_table.GetTable();
+
+    if (parameters.source > -1 && parameters.destination > -1)
+    {
+        for (std::size_t f_counter = 0; f_counter < tfse_table_.size(); f_counter++)
+        {
+            // parameters.source column
+            if (f_counter % result_table.GetNumberOfNodes() == (unsigned long)parameters.source)
+            {
+                tfse_table_[f_counter] = std::numeric_limits<EdgeWeight>::max();
```
Comment (danpat) [inline src/engine/plugins/trip.cpp]: Given that the FI and BruteForce searches are looking for `INVALID_EDGE_WEIGHT`, that's what you should set here, not `::max()`.  Practically, they're the same thing, but if the `INvALID_EDGE_WEIGHT` value changes for some reason, the `if()` logic later on will break.

File: src/engine/plugins/trip.cpp [hunk 2]
```diff
@@ -187,8 +204,90 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+            }
+
+            // parameters.destination row
+            if (f_counter >= (result_table.GetNumberOfNodes() * parameters.destination) &&
+                f_counter < result_table.GetNumberOfNodes() * parameters.destination +
+                                result_table.GetNumberOfNodes())
+            {
+                tfse_table_[f_counter] = std::numeric_limits<EdgeWeight>::max();
```
Comment (danpat) [inline src/engine/plugins/trip.cpp]: Same here.

File: src/engine/plugins/trip.cpp [hunk 3]
```diff
@@ -187,8 +204,90 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+            }
+        }
+
+        tfse_table_[parameters.source * result_table.GetNumberOfNodes() + parameters.source] = 0;
+        tfse_table_[parameters.destination * result_table.GetNumberOfNodes() + parameters.source] =
+            0;
+        tfse_table_[parameters.destination * result_table.GetNumberOfNodes() +
+                    parameters.destination] = 0;
+        tfse_table_[parameters.source * result_table.GetNumberOfNodes() + parameters.destination] =
+            std::numeric_limits<EdgeWeight>::max();
```
Comment (danpat) [inline src/engine/plugins/trip.cpp]: And here.

**Comment by danpat** [comment]
@ghoshkaj For routes >10 waypoints, we will need to make the same logic change to the FarthestInsertion algorithm that got made to BruteForce.  A test for that scenario would be great too.

File: include/engine/api/base_parameters.hpp [hunk 4]
```diff
@@ -96,6 +96,10 @@ struct BaseParameters
                            });
     }
 };
+inline std::ostream &operator<<(std::ostream &out, const BaseParameters &)
+{
+    return out << "BaseParameters()";
```
Comment (danpat) [inline include/engine/api/base_parameters.hpp]: Debugging code?

File: CHANGELOG.md [hunk 5]
```diff
@@ -27,6 +27,8 @@
       - libOSRM now creates an own watcher thread then used in shared memory mode to listen for data updates
     - Tools:
       - Added osrm-extract-conditionals tool for checking conditional values in OSM data
+    - Trip Plugin
+      - Added a new feature that finds the optimal route given a list of location, a start and destination locations. This does not return a roundtrip, but instead gives the optimal route starting at the start and destination locations.
```
Comment (danpat) [inline CHANGELOG.md]: "... optimal route starting at the start and destination locations" - > "... optimal route given fixed start and end points"

File: docs/http.md [hunk 6]
```diff
@@ -329,6 +330,30 @@ In addition to the [general options](#general-options) the following options are
 |geometries  |`polyline` (default), `polyline6`, `geojson`    |Returned route geometry format (influences overview and per step)          |
 |overview    |`simplified` (default), `full`, `false`         |Add overview geometry either full, simplified according to highest zoom level it could be display on, or not at all.|
 
+A second feature of the trip plugin returns the optimal route from a source to a destination while visiting all the other input locations in the middle.
```
Comment (danpat) [inline docs/http.md]: "input locations" -> "waypoints"

File: docs/http.md [hunk 7]
```diff
@@ -329,6 +330,30 @@ In addition to the [general options](#general-options) the following options are
 |geometries  |`polyline` (default), `polyline6`, `geojson`    |Returned route geometry format (influences overview and per step)          |
 |overview    |`simplified` (default), `full`, `false`         |Add overview geometry either full, simplified according to highest zoom level it could be display on, or not at all.|
 
+A second feature of the trip plugin returns the optimal route from a source to a destination while visiting all the other input locations in the middle. 
+As with the roundtrip service, for locations 10 nodes of less, the brute force algorithm is used, while for more than 10 nodes, the farthest insertion heuristic is used.
```
Comment (danpat) [inline docs/http.md]: "10 nodes or less", not "of"

File: docs/http.md [hunk 8]
```diff
@@ -329,6 +330,30 @@ In addition to the [general options](#general-options) the following options are
...
+
+**Example:**
+
+```
+sources=0&destinations=5
+```
+
+|Element     |Values                       |
+|------------|-----------------------------|
+|index       |`0 <= integer < #locations`  |
```
Comment (danpat) [inline docs/http.md]: either `coordinates` or `waypoints`, not `locations`

File: features/testbot/trip.feature [hunk 9]
```diff
@@ -48,12 +47,130 @@ Feature: Basic trip planning
...
+            | bc    |
+
+        When I plan a trip I should get
+            |  waypoints  | status        | message                                                                                  |
+            |  a,b,c      | InvalidInputs | Source or destination indices are greater number of coordinates provided. |
+
+
+    Scenario: Testbot - Trip planning with illegal source
+        Given the query options
+            | source        | 10     |
```
Comment (danpat) [inline features/testbot/trip.feature]: `source` and `destination` should be columns in the "When I plan a trip I should get" columns, not a separate "Given", allowing per-query adjustment.  This will allow some consolidation of these test cases.

File: docs/http.md [hunk 10]
```diff
@@ -311,7 +311,7 @@ All other properties might be undefined.
 
 ### Trip service
 
-The trip plugin solves the Traveling Salesman Problem using a greedy heuristic (farthest-insertion algorithm).
+The trip plugin solves the Traveling Salesman Problem using a greedy heuristic (farthest-insertion algorithm) for more than ten nodes and uses brute force for 10 nodes or less.
```
Comment (daniel-j-h) [inline docs/http.md]: That's not true: < 10 is brute force

https://github.com/Project-OSRM/osrm-backend/blob/b3ef27d104787af8c82ee11d51769a2164bce9b2/src/engine/plugins/trip.cpp#L209

File: docs/http.md [hunk 11]
```diff
@@ -329,6 +329,30 @@ In addition to the [general options](#general-options) the following options are
 |geometries  |`polyline` (default), `polyline6`, `geojson`    |Returned route geometry format (influences overview and per step)          |
 |overview    |`simplified` (default), `full`, `false`         |Add overview geometry either full, simplified according to highest zoom level it could be display on, or not at all.|
 
+A second feature of the trip plugin returns the route from a source to a destination while visiting all the other waypoints in the middle. 
+As with the roundtrip service, for fewer than 10 waypoints, the brute force algorithm is used, while for more than 10 waypoints the farthest insertion heuristic is used.
```
Comment (daniel-j-h) [inline docs/http.md]: Same here:
- < 10 brute force
- \>= 10 farthest insertion

File: docs/http.md [hunk 12]
```diff
@@ -329,6 +329,30 @@ In addition to the [general options](#general-options) the following options are
...
+
+|Option      |Values                                          |Description                                                                                                                              |
+|------------|------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|
+|source      |`{index of coordinate in coordinates provided}` |Will error if destionation is not provided as well. Returns trip from source to destination routing through all other locations provided.|
```
Comment (daniel-j-h) [inline docs/http.md]: destination - grammar

File: features/step_definitions/trip.js [hunk 13]
```diff
@@ -35,11 +34,12 @@ module.exports = function () {
                     }
 
                     if (headers.has('status')) {
-                        got.status = json.status.toString();
+                        got.status = json.code;
```
Comment (daniel-j-h) [inline features/step_definitions/trip.js]: Why the change from the status message to the error code here?

File: features/testbot/trip.feature [hunk 14]
```diff
@@ -48,12 +47,130 @@ Feature: Basic trip planning
...
+            """
+
+        And the ways
+            | nodes |
+            | ab    |
+            | bc    |
+
+        When I plan a trip I should get
+            |  waypoints  | status        | message                                                                                  |
+            |  a,b,c      | InvalidInputs | Source or destination indices are greater number of coordinates provided. |
```
Comment (daniel-j-h) [inline features/testbot/trip.feature]: All these tests should be unit tests - check the unit tests dir and the library tests there, e.g. for route.

File: include/engine/api/trip_parameters.hpp [hunk 15]
```diff
@@ -47,7 +47,27 @@ namespace api
  */
 struct TripParameters : public RouteParameters
 {
-    // bool IsValid() const; Falls back to base class
```
Comment (daniel-j-h) [inline include/engine/api/trip_parameters.hpp]: remove

File: include/engine/api/trip_parameters.hpp [hunk 16]
```diff
@@ -47,7 +47,27 @@ namespace api
...
+    TripParameters()
+        : RouteParameters(false,
+                          false,
+                          false,
+                          RouteParameters::GeometriesType::Polyline,
+                          RouteParameters::OverviewType::Simplified,
+                          {}),
```
Comment (daniel-j-h) [inline include/engine/api/trip_parameters.hpp]: Why setting parameters here for RouteParameters?

Can't this be `TripParameters() = default;` and setting `source = -1` and `destination = -1` below?

File: include/engine/api/trip_parameters.hpp [hunk 17]
```diff
@@ -47,7 +47,27 @@ namespace api
...
+    }
+
+    template <typename... Args>
+    TripParameters(int source_, int destination_, Args &&... args_)
+        : RouteParameters{std::forward<Args>(args_)...}, source{source_}, destination{destination_}
+    {
+    }
+
+    int source;
+    int destination;
```
Comment (daniel-j-h) [inline include/engine/api/trip_parameters.hpp]: Source and destination are indices - they should really be `std::size_t` and maybe even wrapper in an `optional<T>`

File: include/engine/api/trip_parameters.hpp [hunk 18]
```diff
@@ -47,7 +47,27 @@ namespace api
...
+
+    bool IsValid() const { return RouteParameters::IsValid() && source >= -1 && destination >= -1; }
```
Comment (daniel-j-h) [inline include/engine/api/trip_parameters.hpp]: You check >= -1 but what about < coordinates size?

File: include/engine/trip/trip_brute_force.hpp [hunk 19]
```diff
@@ -31,7 +31,20 @@ EdgeWeight ReturnDistance(const util::DistTableWrapper<EdgeWeight> &dist_table,
...
+        // Fixed start and end trip needs to escape this path if a max edge weight is found on the
+        // dist table
+        if (dist_table(location_order[i], location_order[(i + 1) % component_size]) ==
+            INVALID_EDGE_WEIGHT)
+        {
+            return INVALID_EDGE_WEIGHT;
+        }
+        else
+        {
+            route_dist += dist_table(location_order[i], location_order[(i + 1) % component_size]);
```
Comment (daniel-j-h) [inline include/engine/trip/trip_brute_force.hpp]: Call dist_table once outside - then store result and make the check

File: include/engine/trip/trip_brute_force.hpp [hunk 20]
```diff
@@ -31,7 +31,20 @@ EdgeWeight ReturnDistance(const util::DistTableWrapper<EdgeWeight> &dist_table,
     std::size_t i = 0;
     while (i < location_order.size() && (route_dist < min_route_dist))
     {
-        route_dist += dist_table(location_order[i], location_order[(i + 1) % component_size]);
+        // Check for path that touches INVALID_EDGE_WEIGHT (which is MAX_32INT)
+        // Fixed start and end trip needs to escape this path if a max edge weight is found on the
+        // dist table
```
Comment (daniel-j-h) [inline include/engine/trip/trip_brute_force.hpp]: Hum.. why?

File: include/engine/trip/trip_farthest_insertion.hpp [hunk 21]
```diff
@@ -50,8 +50,8 @@ GetShortestRoundTrip(const NodeID new_loc,
         const auto dist_to = dist_table(new_loc, *to_node);
         const auto trip_dist = dist_from + dist_to - dist_table(*from_node, *to_node);
 
-        BOOST_ASSERT_MSG(dist_from != INVALID_EDGE_WEIGHT, "distance has invalid edge weight");
-        BOOST_ASSERT_MSG(dist_to != INVALID_EDGE_WEIGHT, "distance has invalid edge weight");
+        if (dist_from == INVALID_EDGE_WEIGHT || dist_to == INVALID_EDGE_WEIGHT)
+            continue;
```
Comment (daniel-j-h) [inline include/engine/trip/trip_farthest_insertion.hpp]: Same here, what's going on here?

File: include/util/dist_table_wrapper.hpp [hunk 22]
```diff
@@ -46,6 +46,30 @@ template <typename T> class DistTableWrapper
         return table_[index];
     }
 
+    void InvalidateRoute(NodeID from, NodeID to)
```
Comment (daniel-j-h) [inline include/util/dist_table_wrapper.hpp]: The wrapper holds `T`s so you probably should accept `T`s to index the vector.

File: src/engine/plugins/trip.cpp [hunk 23]
```diff
@@ -160,6 +176,14 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
         return Error("InvalidValue", "Invalid coordinate value.", json_result);
     }
 
+    if (parameters.source >= static_cast<int>(parameters.coordinates.size()) ||
+        parameters.destination >= static_cast<int>(parameters.coordinates.size()))
+    {
+        return Error("InvalidInputs",
```
Comment (daniel-j-h) [inline src/engine/plugins/trip.cpp]: Not a documented v5 error code.

File: src/engine/plugins/trip.cpp [hunk 24]
```diff
@@ -187,8 +211,76 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+    // d 34 30 18 0  15
+    // e 30 34 32 15 0
+
+    // New Table
+    //   a        b         c        d         e
+    // a 0        15        10000    34        30
+    // b 10000    0         25       30        34
+    // c 0        10000     0        10000     10000
+    // d 10000    30        18       0         15
+    // e 10000    34        32       15        0
```
Comment (daniel-j-h) [inline src/engine/plugins/trip.cpp]: So what does it do?

File: src/engine/plugins/trip.cpp [hunk 25]
```diff
@@ -187,8 +211,76 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+
+    if (parameters.source > -1 && parameters.destination > -1)
+    {
+        // parameters.source column
+        for (NodeID i = 0; i < result_table.GetNumberOfNodes(); i++) {
+            if (i == (NodeID)parameters.source) continue;
```
Comment (daniel-j-h) [inline src/engine/plugins/trip.cpp]: format all files you're touching

File: unit_tests/server/parameters_parser.cpp [hunk 26]
```diff
@@ -439,9 +439,19 @@ BOOST_AUTO_TEST_CASE(valid_trip_urls)
...
+    TripParameters reference_2{};
+    reference_2.coordinates = coords_1;
+    reference_2.source = 1;
+    reference_2.destination = 2;
+    auto result_2 = parseParameters<TripParameters>("1,2;3,4?source=1&destination=2");
+    BOOST_CHECK(result_2);
+    BOOST_CHECK_EQUAL(reference_2.source, result_2->source);
+    BOOST_CHECK_EQUAL(reference_2.destination, result_2->destination);
+    CHECK_EQUAL_RANGE(reference_2.radiuses, result_2->radiuses);
+    CHECK_EQUAL_RANGE(reference_2.coordinates, result_2->coordinates);
```
Comment (daniel-j-h) [inline unit_tests/server/parameters_parser.cpp]: More tests please. For parsing but also library tests for responses.

File: include/util/dist_table_wrapper.hpp [hunk 27]
```diff
@@ -46,6 +46,30 @@ template <typename T> class DistTableWrapper
...
+        BOOST_ASSERT_MSG(to < number_of_nodes_, "to ID is out of bound");
+
+        const auto index = from * number_of_nodes_ + to;
+
+        BOOST_ASSERT_MSG(index < table_.size(), "index is out of bound");
+
+        table_[index] = INVALID_EDGE_WEIGHT;
+    }
+
+    void ShortcutRoute(NodeID from, NodeID to)
```
Comment (chaupow) [inline include/util/dist_table_wrapper.hpp]: I'm not totally convinced that `ShortcutRoute ` and `InvalidateRoute ` are good names. If you want to keep them more general as they are now, can you add a comment why we do this here? so something like
`force the roundtrip to start at source and end at destination by virtually squashing them together`

or you can maybe name them `squashSourceAndDestination` and `separateDestination`

File: src/engine/plugins/trip.cpp [hunk 28]
```diff
@@ -187,8 +211,76 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
+
+        result_table.ShortcutRoute(parameters.destination, parameters.source);
+        result_table.InvalidateRoute(parameters.source, parameters.destination);
+    }
+
     // get scc components
-    SCC_Component scc = SplitUnaccessibleLocations(number_of_locations, result_table);
+    SCC_Component scc = SplitUnaccessibleLocations(result_table.GetNumberOfNodes(), result_table);
+
+    if (parameters.source > -1 && parameters.destination > -1) // check if fixed start and end
```
Comment (chaupow) [inline src/engine/plugins/trip.cpp]: later you're checking this to set  `bool roundtrip`. you could also set the boolean already here and use it.

File: include/engine/api/trip_parameters.hpp (hunk 15 above)
Comment (ghoshkaj) [inline include/engine/api/trip_parameters.hpp]: I think I _did_ remove this :)

File: features/step_definitions/trip.js (hunk 13 above)
Comment (ghoshkaj) [inline features/step_definitions/trip.js]: The status.toString() code was copy pasted from another cucumber test but wasn't really being used in the trip context (it was broken code essentially with no way to trigger it to check), so I changed it so that it does work.

File: include/util/dist_table_wrapper.hpp (hunk 22 above)
Comment (ghoshkaj) [inline include/util/dist_table_wrapper.hpp]: The DistTableWrapper is a wrapper around a vector that holds <T>, in the trip case the <T> is an EdgeWeight. This method here simply accepts NodeIDs to access the correct position of the necessary EdgeWeight in the table vector of EdgeWeights. So keeping the function as-is should be fine in this case.

File: src/engine/plugins/trip.cpp [hunk 29]
```diff
@@ -117,24 +117,40 @@ SCC_Component SplitUnaccessibleLocations(const std::size_t number_of_locations,
...
+
+    if (roundtrip)
+    {
+        BOOST_ASSERT(min_route.segment_end_coordinates.size() == trip.size());
+    }
+    else
+    {
+        // trip comes out to be something like 0 1 4 3 2, so the sizes don't
+        BOOST_ASSERT(min_route.segment_end_coordinates.size() == trip.size() - 1);
+    }
```
Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: I did this, but it got a bit confusing for `&&` / `||` noobs like me to keep the logic straight. So I reverted back to the if statements for clarity and simplicity's sake.

**Comment by daniel-j-h** [comment]
Tasks for `node-osrm` ticketed here: https://github.com/Project-OSRM/node-osrm/issues/294.

**Comment by ghoshkaj** [comment]
To Do (@chaupow if you want to get started on any of these)
- Figure out what to do for scc.component > 1 and fix error handling accordingly
- Add `roundtrip` parameter parser test(s)
- Add the `source={fixed|any}` and `destination={fixed|any}` parameters
- Add parameter parser tests, unit tests and cucumber tests for the scenarios involving fixed start and fixed end parameters

File: features/testbot/trip.feature [hunk 30]
```diff
@@ -48,12 +65,85 @@ Feature: Basic trip planning
...
+            | fg    |
+            | gh    |
+            | hi    |
+            | ij    |
+            | jk    |
+            | kl    |
+            | la    |
+
+        When I plan a trip I should get
+            | waypoints               | trips         | source |
```
Comment (TheMarex) [inline features/testbot/trip.feature]: Why is `source` a result column? This should be a query parameter like in https://github.com/Project-OSRM/osrm-backend/pull/3408/files#diff-19534b2b92b1d620a3a6390c0f0b378dR100 no?

File: features/testbot/trip.feature [hunk 31]
```diff
@@ -85,13 +175,106 @@ Feature: Basic trip planning
...
+            """
+
+         And the ways
+            | nodes |
+            | ab    |
+            | dc    |
+
+         When I plan a trip I should get
+            |  waypoints    | source  | destination | roundtrip | status           | message                                       |
+            |  a,b,c,d      | first   | last        | false     | NoTrips          | No trip visiting all destinations possible.   |
```
Comment (TheMarex) [inline features/testbot/trip.feature]: Same here: `source` and `destination` should be query parameters not part of the result table.

File: features/testbot/trip.feature [hunk 32]
```diff
@@ -85,13 +175,106 @@ Feature: Basic trip planning
...
+            | be    |
+            | cd    |
+            | ce    |
+            | de    |
 
         When I plan a trip I should get
-            | waypoints                       | trips               |
-            | a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p | defghijklabcd,mnopm |
+            |  waypoints  | source | destination |roundtrip | trips  | durations         | distance               |
+            |  a,b,d,e,c  | first  | last        | false    | abedc  | 8.200000000000001 | 81.6                   |
```
Comment (TheMarex) [inline features/testbot/trip.feature]: Same here.

File: features/testbot/trip.feature [hunk 33]
```diff
@@ -85,13 +175,106 @@ Feature: Basic trip planning
...
+            | de    |
+            | ef    |
+            | fg    |
+            | gh    |
+            | hi    |
+            | ij    |
+            | jk    |
+
+        When I plan a trip I should get
+            |  waypoints              | source | destination | roundtrip |  trips       | durations  | distance  |
```
Comment (TheMarex) [inline features/testbot/trip.feature]: Same here.

File: include/engine/trip/trip_brute_force.hpp [hunk 34]
```diff
@@ -25,62 +26,77 @@ namespace trip
...
 {
     EdgeWeight route_dist = 0;
     std::size_t i = 0;
     while (i < location_order.size() && (route_dist < min_route_dist))
     {
-        route_dist += dist_table(location_order[i], location_order[(i + 1) % component_size]);
-        BOOST_ASSERT_MSG(dist_table(location_order[i], location_order[(i + 1) % component_size]) !=
-                             INVALID_EDGE_WEIGHT,
-                         "invalid route found");
+
```
Comment (TheMarex) [inline include/engine/trip/trip_brute_force.hpp]: A better naming for `i` would could be `current_index` and for `next_index = (current_index + 1) % number_of_locations`.

File: include/engine/trip/trip_farthest_insertion.hpp [hunk 35]
```diff
@@ -158,57 +153,26 @@ std::vector<NodeID> FarthestInsertionTrip(const NodeIDIterator &start,
...
-                    max_from = *x;
-                    max_to = *y;
-                }
-            }
-        }
-    }
+    // find the pair of location with the biggest distance and make the pair the initial start
+    // trip. Skipping over the very first element (0,0), we make sure not to end up with the
+    // same start/end in the special case where all entries are the same.
+    const auto index = std::distance(
```
Comment (TheMarex) [inline include/engine/trip/trip_farthest_insertion.hpp]: `index` should be named more precisely: `weight_index`.

File: include/engine/trip/trip_farthest_insertion.hpp [hunk 36]
```diff
@@ -158,57 +153,26 @@ std::vector<NodeID> FarthestInsertionTrip(const NodeIDIterator &start,
...
+        std::begin(dist_table), std::max_element(std::begin(dist_table) + 1, std::end(dist_table)));
+    NodeID max_from = index / number_of_locations;
```
Comment (TheMarex) [inline include/engine/trip/trip_farthest_insertion.hpp]: Needs a comment for computing the row from the `weight_index`.

File: include/engine/trip/trip_farthest_insertion.hpp [hunk 37]
```diff
@@ -158,57 +153,26 @@ std::vector<NodeID> FarthestInsertionTrip(const NodeIDIterator &start,
...
+    NodeID max_to = index % number_of_locations;
```
Comment (TheMarex) [inline include/engine/trip/trip_farthest_insertion.hpp]: Needs a comment for computing the column from the `weight_index`.

File: include/util/dist_table_wrapper.hpp (hunk 22 above)
Comment (TheMarex) [inline include/util/dist_table_wrapper.hpp]: Should these functions be part of the `DistTableWrapper`? This class seems very generic and the functions are only relevant in the context of the `trip` plugin.

Another way to do it would be to have two functions `invalidateRoute(DistanceTableWrapper &dist, NodeID from, NodeID to)` and `shortcutRoute(DistanceTableWrapper &dist, NodeID from, NodeID to)` in the trip plugin.

File: include/util/typedefs.hpp [hunk 38]
```diff
@@ -60,6 +60,8 @@ using NameID = std::uint32_t;
 using EdgeWeight = std::int32_t;
 using TurnPenalty = std::int16_t; // turn penalty in 100ms units
 
+static const std::size_t INVALID_SIZE_T = std::numeric_limits<std::size_t>::max();
```
Comment (TheMarex) [inline include/util/typedefs.hpp]: `INVALID_INDEX` would probably be a more fitting name, since that is what it is used to denote in the using code.

File: src/engine/plugins/trip.cpp [hunk 39]
```diff
@@ -28,129 +27,151 @@ namespace engine
...
+        source_id = 0;
+    }
+    if (parameters.destination == api::TripParameters::DestinationType::Last)
+    {
+        BOOST_ASSERT(number_of_locations > 0);
+        destination_id = number_of_locations - 1;
+    }
+    bool fixed_start = (source_id == 0);
+    bool fixed_end = (destination_id == number_of_locations - 1);
+    bool fixed_start_and_end = fixed_start && fixed_end;
```
Comment (TheMarex) [inline src/engine/plugins/trip.cpp]: It might be less confusing to not pre-compute `fixed_start_and_end`, because in the code below it was not clear if this is somehow different then `fixed_start && fixed_end`.

File: src/engine/plugins/trip.cpp [hunk 40]
```diff
@@ -187,56 +214,48 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
-        return Error("NoTrips", "Cannot find trips", json_result);
+        trip = trip::FarthestInsertionTrip(number_of_locations, result_table);
     }
 
-    // compute all round trip routes
-    std::vector<InternalRouteResult> routes;
-    routes.reserve(trips.size());
-    for (const auto &trip : trips)
+    // rotate result such that roundtrip starts at node with index 0
+    if (!(fixed_end && !fixed_start_and_end))
```
Comment (TheMarex) [inline src/engine/plugins/trip.cpp]: Same as `!fixed_end || fixed_start_and_end` == `!fixed_end && fixed_start || !fixed_end && fixed_end` == `fixed_start && !fixed_end`

File: src/engine/plugins/trip.cpp [hunk 41]
```diff
@@ -187,56 +214,48 @@ Status TripPlugin::HandleRequest(const std::shared_ptr<const datafacade::BaseDat
...
     {
-        routes.push_back(ComputeRoute(facade, snapped_phantoms, trip));
+        auto desired_start_index = std::find(std::begin(trip), std::end(trip), 0);
+        BOOST_ASSERT(desired_start_index != std::end(trip));
+        std::rotate(std::begin(trip), desired_start_index, std::end(trip));
     }
+    else if (fixed_end && !fixed_start_and_end && parameters.roundtrip)
```
Comment (TheMarex) [inline src/engine/plugins/trip.cpp]: Same as `fixed_end && !fixed_start && parameters.roundtrip`

File: features/testbot/trip.feature (hunk 30 above)
Comment (danpat) [inline features/testbot/trip.feature]: @TheMarex I asked @ghoshkaj to move `source` and `dest` into the Results table - they're query parameters in the same way that `waypoints` are.

Having these in the `When` table allows multiple combinations for the same scenario - putting them in the `Given the query options` clause limits us to one set per scenario.

Comment (danpat) [inline features/testbot/trip.feature]: Note that there is precedence for this style in the `testbot/matching.feature` where we supply timestamps.

Comment (chaupow) [inline features/testbot/trip.feature]: @TheMarex @danpat: and now... - fight! 😄
kidding. I actually agree with @danpat but in most of these cases we only test one parameter combination so the argument is just semi-valid.

If you both don't mind and there are no further opinions on these, I'd leave it as it is just because it's less work.

File: include/util/typedefs.hpp (hunk 38 above)
Comment (chaupow) [inline include/util/typedefs.hpp]: HAHAHA. True. Thanks. invalid_size_t is wrong indeed

File: features/testbot/trip.feature (hunk 31 above)
Comment (ghoshkaj) [inline features/testbot/trip.feature]: Leaving this as is due to comment above.

File: features/testbot/trip.feature (hunk 32 above)
Comment (ghoshkaj) [inline features/testbot/trip.feature]: Leaving this as is due to comment above.

File: features/testbot/trip.feature (hunk 33 above)
Comment (ghoshkaj) [inline features/testbot/trip.feature]: Leaving this as is due to comment above.

File: src/engine/plugins/trip.cpp (hunk 40 above)
Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: `!(fixed_end && !fixed_start_and_end` == `!fixed_end || fixed_start_and_end` == `!fixed_end || (fixed_start && fixed_end)` == `(!fixed_end || fixed_start) && (!fixed_end || fixed_end)` == `!fixed_end || fixed_start` == `fixed_start`

Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: No `!fixed_end || fixed_start_and_end == !fixed_end || (fixed_start && fixed_end)` is wrong from my comment above. Ugh it's so late, I can't keep the logic straight :/ 

In any case, using `fixed_start && !fixed_end` from Patrick's comment makes a test fail :/

Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: While using `fixed_start` lets them all pass.

Comment (oxidase) [inline src/engine/plugins/trip.cpp]: hm,  can it be simplified? There are only 16 2-variable boolean functions

Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: Okay no matter how many times I work it out, I always get to `!fixed_end || fixed_start` and this lets the tests pass happily so this is what I'll keep it as.

Comment (ghoshkaj) [inline src/engine/plugins/trip.cpp]: Realistically, what this case is taking care of are the following scenarios:
`!fixed_end` OR
`fixed_start` OR
`fixed_start && fixed_end`
so I'll either write all these cases out or comment with this.

Comment (oxidase) [inline src/engine/plugins/trip.cpp]: `!fixed_end || fixed_start` is the same as `!fixed_end OR
fixed_start OR fixed_start && fixed_end`, because `fixed_start OR fixed_start && fixed_end` is equal to `fixed_start`
